openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+           The integrator to use for Context creation. The eventual
+           heat bath temperature must be consistent with the
+           thermodynamic state.
+        platform : simtk.openmm.Platform, optional
+           Platform to use. If None, OpenMM tries to select the fastest
+           available platform. Default is None.
+
+        Returns
+        -------
+        context : simtk.openmm.Context
+           The created OpenMM Context object.
+
+        Raises
+        ------
+        ThermodynamicsError
+            If the integrator has a temperature different from this
+            ThermodynamicState.
+
+        Examples
+        --------
+        Creating a context with a thermostated integrator means that
+        the context system will not have a thermostat.
+
+        >>> from simtk import openmm, unit
+        >>> from openmmtools import testsystems
+        >>> toluene = testsystems.TolueneVacuum()
+        >>> state = ThermodynamicState(toluene.system, 300*unit.kelvin)
+        >>> integrator = openmm.VerletIntegrator(1.0*unit.femtosecond)
+        >>> context = state.create_context(integrator)
+        >>> system = context.getSystem()
+        >>> [force.__class__.__name__ for force in system.getForces()
+        ...  if 'Thermostat' in force.__class__.__name__]
+        ['AndersenThermostat']
+
+        The thermostat is removed if we choose an integrator coupled
+        to a heat bath.
+
+        >>> integrator = openmm.LangevinIntegrator(300*unit.kelvin, 5.0/unit.picosecond,
+        ...                                        2.0*unit.femtosecond)
+        >>> context = state.create_context(integrator)
+        >>> [force.__class__.__name__ for force in system.getForces()
+        ...  if 'Thermostat' in force.__class__.__name__]
+        []
+
+        """
+        # Check that integrator is consistent and if it is thermostated.
+        # With CompoundIntegrator, at least one must be thermostated.
+        is_thermostated = False
+        if isinstance(integrator, openmm.CompoundIntegrator):
+            for integrator_id in range(integrator.getNumIntegrators()):
+                _integrator = integrator.getIntegrator(integrator_id)
+                is_thermostated = is_thermostated or self._is_integrator_thermostated(_integrator)
+        else:
+            is_thermostated = self._is_integrator_thermostated(integrator)
+
+        # If integrator is coupled to heat bath, remove system thermostat.
+        system = copy.deepcopy(self._system)
+        if is_thermostated:
+            self._remove_thermostat(system)
+
+        # Create platform.
+        if platform is None:
+            return openmm.Context(system, integrator)
+        else:
+            return openmm.Context(system, integrator, platform)
+
+    def apply_to_context(self, context):
+        """Apply this ThermodynamicState to the context.
+
+        The method apply_to_context does *not* check for the compatibility
```

Comment:
We would need to apply the `is_context_compatible` here since `ThermodynamicState` doesn't know which state created this context, but that function is expensive because it serializes the system. I thought it would be preferable to leave this responsibility to the user of the class who already knows which state created which context (and can call `is_state_compatible` which is way cheaper).